### PR TITLE
Always create `Delayed` object for `store`

### PR DESCRIPTION
**Pull request by jakirkham**
Pulls some changes out of PR ( https://github.com/dask/dask/pull/2980 ). 

Instead of handling the `compute=True` and `compute=False` cases of `store` through two different mechanisms, always create a `Delayed` object at the end. Then either call `compute` on the object if `compute=True` or return the `Delayed` object if `compute=False`.

### CODE REVIEW

**Comment by mrocklin** [comment]
Seems fine to me

**Comment by jakirkham** [comment]
Thanks for taking a look. Wasn't sure if there was any advantage to using `compute_as_if_collection` compared to calling `compute` on a `Delayed` object. If not, seems like using the same strategy for both would be nicer.

Failure is due to issue ( https://github.com/dask/dask/issues/2996 ). So should be cleared up once that is resolved.

**Comment by jakirkham** [comment]
Planning on merging tomorrow if there are no other comments.

File: dask/array/core.py [hunk 1]
```diff
@@ -894,17 +896,17 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+        update.update(src.dask)
         update.update(dsk)
         updates.update(update)
 
     name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge((name, updates), *[src.dask for src in sources])
+    dsk = sharedict.merge({name: keys}, updates)
+    result = Delayed(name, dsk)
     if compute:
-        compute_as_if_collection(Array, dsk, keys, **kwargs)
```
Comment (mrocklin) [inline dask/array/core.py]: OK, we probably needed this line.  I apologize for not catching this earlier, but the reason we did this before was to properly engage all of the dask.array optimization pipeline.  We lose that by just passing around the task graph.

File: dask/array/core.py [hunk 2]
```diff
@@ -894,17 +896,17 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
...
+        result.compute()
     else:
-        from ..delayed import Delayed
-        dsk.update({name: keys})
-        return Delayed(name, dsk)
```
Comment (mrocklin) [inline dask/array/core.py]: This code down here was likely also flawed previously.  We might call the optimize machinery as we do in `Array.to_delayed`

```python
    def to_delayed(self):
        """ Convert Array into dask Delayed objects

        Returns an array of values, one value per chunk.

        See Also
        --------
        dask.array.from_delayed
        """
        from ..delayed import Delayed
        keys = self.__dask_keys__()
        dsk = self.__dask_optimize__(self.__dask_graph__(), keys)
        L = ndeepmap(self.ndim, lambda k: Delayed(k, dsk), keys)
        return np.array(L, dtype=object)
```

**Comment by jakirkham** [comment]
Ok. Happy to revert. That said, I’m away from the computer currently (on mobile sans `git`).

**Comment by mrocklin** [comment]
No big rush, we've been doing the compute=False approach here badly for a
long while.  I would just roll in changes to the larger PR.

**Comment by jakirkham** [comment]
Ok, I'll assign myself so I don't forget.

**Comment by jakirkham** [comment]
Have added PR ( https://github.com/dask/dask/pull/3005 ) to revert this.

**Pull request merged**